toolbox/relnotes/main.go

```diff
@@ -0,0 +1,784 @@
+		}
+	}
+
+	for _, file := range files {
+		fn := filepath.Base(file)
+		sha, err := u.GetSha256(file)
+		if err != nil {
+			return fmt.Errorf("failed to calc SHA256 of file %s: %s", file, err)
+		}
+		f.WriteString(fmt.Sprintf("[%s](%s/%s/%s) | `%s`\n", fn, urlPrefix, releaseTag, fn, sha))
+	}
+	return nil
+}
+
+// minorReleases performs a minor (vX.Y.0) release by fetching the release template and aggregate
+// previous release in series.
+func minorRelease(f *os.File, release, draftURL, changelogURL string) {
+
+	// Check for draft and use it if available
+	log.Printf("Checking if draft release notes exist for %s...", release)
+
+	resp, err := http.Get(draftURL)
+	if err == nil {
+		defer resp.Body.Close()
+	}
+
+	if err == nil && resp.StatusCode == 200 {
+		log.Printf("Draft found - using for release notes...")
+		_, err = io.Copy(f, resp.Body)
+		if err != nil {
+			log.Printf("error during copy to file: %s", err)
+			return
+		}
+		f.WriteString("\n")
+	} else {
+		log.Printf("Failed to find draft - creating generic template... (error message/status code printed below)")
+		if err != nil {
+			log.Printf("Error message: %s", err)
+		} else {
+			log.Printf("Response status code: %d", resp.StatusCode)
+		}
+		f.WriteString("## Major Themes\n\n* TBD\n\n## Other notable improvements\n\n* TBD\n\n## Known Issues\n\n* TBD\n\n## Provider-specific Notes\n\n* TBD\n\n")
+	}
+
+	// Aggregate all previous release in series
+	f.WriteString(fmt.Sprintf("### Previous Release Included in %s\n\n", release))
+
+	// Regexp Example:
+	// Assume the release tag is v1.7.0, this regexp matches "- [v1.7.0-" in
+	//     "- [v1.7.0-rc.1](#v170-rc1)"
+	//     "- [v1.7.0-beta.2](#v170-beta2)"
+	reAnchor, _ := regexp.Compile(fmt.Sprintf("- \\[%s-", release))
+
+	resp, err = http.Get(changelogURL)
+	if err == nil {
+		defer resp.Body.Close()
+	}
+
+	if err == nil && resp.StatusCode == 200 {
+		buf := new(bytes.Buffer)
+		buf.ReadFrom(resp.Body)
+		for _, line := range strings.Split(buf.String(), "\n") {
+			if anchor := reAnchor.FindStringSubmatch(line); anchor != nil {
+				f.WriteString(line + "\n")
+			}
+		}
+		f.WriteString("\n")
+	} else {
+		log.Printf("Failed to fetch past changelog for minor release - continuing... (error message/status code printed below)")
+		if err != nil {
+			log.Printf("Error message: %s", err)
+		} else {
+			log.Printf("Response status code: %d", resp.StatusCode)
+		}
+	}
+}
+
+// patchRelease performs a patch (vX.Y.Z) release by printing out all the related changes.
+func patchRelease(f *os.File, info *ReleaseInfo) {
+	// Release note for different labels
+	f.WriteString(fmt.Sprintf("## Changelog since %s\n\n", info.startTag))
+
+	if len(info.releaseActionRequiredPRs) > 0 {
+		f.WriteString("### Action Required\n\n")
+		for _, pr := range info.releaseActionRequiredPRs {
+			f.WriteString(fmt.Sprintf("* %s (#%d, @%s)\n", extractReleaseNoteFromPR(info.prMap[pr]), pr, *info.prMap[pr].User.Login))
+		}
+		f.WriteString("\n")
+	}
+
```

Comment:
I also see in the CHANGELOG we have a section `Deprecations`. How is it generated? I guess we may want a new `release-note-deprecation` ?

Reply:
I think it's from the release note drafts in kubernetes/features repo. I don't know how are the drafts generated though @david-mcmahon

Reply:
Release Notes for new minor (x.y.0) releases are populated from their respective feature branch repos, such as https://github.com/kubernetes/features/blob/master/release-1.7/release-notes-draft.md.  These files are updated throughout the release cycle leading up to the .0 release.

Reply:
A new label `release-note-deprecation` will be helpful to auto-generate the section `Deprecations`. The items in this section are from the `release-note-action-required` PRs and then trriaged to this section manually I guess.